crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,436 @@
+                Ok(())
+            }
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    fn commitment(&self, version: SnapshotVersion) -> Result<Self::Commitment, StakeTableError> {
+        match version {
+            // IMPORTANT: we don't support committing the head version b/c it's not finalized.
+            SnapshotVersion::Head => Err(StakeTableError::SnapshotUnsupported),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_comm),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_comm),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn total_stake(&self, version: SnapshotVersion) -> Result<Self::Amount, StakeTableError> {
+        match version {
+            SnapshotVersion::Head => Ok(self.head_total_stake),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_total_stake),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_total_stake),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn len(&self, version: SnapshotVersion) -> Result<usize, StakeTableError> {
+        match version {
+            SnapshotVersion::Head => Ok(self.head.len()),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start.len()),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start.len()),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn contains_key(&self, key: &Self::Key) -> bool {
+        self.mapping.contains_key(key)
+    }
+
+    fn lookup(
+        &self,
+        version: SnapshotVersion,
+        key: &Self::Key,
+    ) -> Result<(Self::Amount, Self::LookupProof), StakeTableError> {
+        match self.mapping.get(key) {
+            Some(&pos) => match version {
+                SnapshotVersion::Head => {
+                    if pos >= self.head.len() {
+                        Err(StakeTableError::KeyNotFound)
+                    } else {
+                        Ok((self.head[pos].1, ()))
+                    }
+                }
+                SnapshotVersion::EpochStart => {
+                    if pos >= self.epoch_start.len() {
+                        Err(StakeTableError::KeyNotFound)
+                    } else {
+                        Ok((self.epoch_start[pos].1, ()))
+                    }
+                }
+                SnapshotVersion::LastEpochStart => {
+                    if pos >= self.last_epoch_start.len() {
+                        Err(StakeTableError::KeyNotFound)
+                    } else {
+                        Ok((self.last_epoch_start[pos].1, ()))
+                    }
+                }
+                SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+            },
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    fn simple_lookup(
```

Comment:
What's the difference from lookup instantiation?

Reply:
Originally with Merkle tree design, this is the one w/o a membership proof.

Reply:
In this case where two API implementations are the same, maybe we can reduce code duplications by directly calling `lookup()` here.

Reply:
nope, I meant to use `simple_lookup` for looking up `amount` only, and `lookup` will return together with a membership proof. 

```rust
    /// Returns the stakes withhelded by a public key, None if the key is not registered.
    /// If you need a lookup proof, use [`Self::lookup()`] instead (which is usually more expensive).
    ///
    /// # Errors
    /// Return err if the `version` is not supported or `key` doesn't exist.
    fn simple_lookup(
        &self,
        version: SnapshotVersion,
        key: &Self::Key,
    ) -> Result<Self::Amount, StakeTableError>;
```

meanwhile, I agree with Binyi that we can reduce duplication whenever possible.